### ISSUE TITLE
Provide better visibility into conflicting versions

### ISSUE THREAD

**Issue by kardianos** [kardianos/vendor#15]
Context: http://www.reddit.com/r/golang/comments/390yes/gb_a_project_based_build_tool_for_the_go/cs2clg0?context=3

I want to find or create sane tooling that deals with the case of my project wanting to vendor a project which itself vendors other projects. For example I want to vendor A at version a1, and B at version b1, but B@b1 already vendors A at a2.
"Versions" here might be "head", a specific git branch, a specific git tag, or even a specific git sha. It depends on how much I trust my deps and that should be up to me.
Most importantly, the tooling has to be human friendly. Show me what you're doing. Explain why. When you hit a conflict, help me figure out WTF to do - don't just barf. In the end it's humans who manage the deps and make decisions about them.
I reject the notion that libraries shouldn't vendor deps - it just doesn't match reality.
I have been bouncing a lot of ideas around about how to do it, and I admit it gets complicated. Consider a project A that has two deps trees ("->" means "depends on")
A -> B/x@b1 -> C@c1 -> D@d1 -> C/y@c1
A -> E@e1 -> C@c2 -> D@d1 -> C/y@c2
If you think of these vendoring ops sequentially in time:
A imports B/x at explicit version b1. B depends on C at explicit version c1. C depends on D with no version so we lock version d1. D depends on C/y with no version so we lock version c1, which we already have.
A then imports E at explicit version e1. E depends on C at explicit version c2, so we rename it to C.c2 locally. C.c2 depends on D with no version so we lock version d1. D depends on C/y with no version so we ... ?
a) lock to C@c1 - this is wrong because we used C.c2 up-graph. b) lock to C.c2 - this is right, but it requires a fork of D@d1 for each different version of C.
Basically it amounts to recursive vendoring. Alternate realities, as it were, for each node in the graph that explicitly vendors (which I would expect any non-trivial project to do if the tooling is good).
This stems from an actual situation we find ourselves in.

**Comment by kardianos** [comment]
I agree with you that tools should provide insight and not just barf.

I agree with you that libraries should be able to vendor deps.

Right now the state of this tool is that it brings deps of deps to the top level and chooses a winner. What it doesn't do is try to resolve versions. There are a few sub-issues that this issue brings up:
1) Inspect versions when bringing dependencies forward.
2) Provide some type of easily grokable dependency graph (with versions) that can be used when resolving dependencies. Allow filtering of nodes and forward only and backward only modes.

**Comment by gedw99** [comment]
@kardianos +1: A dependency graph would help allot for large projects. as things get bigger they get move complex. Any thoughts on what web based tool to use for dependency graph ? I have been thinking about the same thing myself and am using a QML based tool for this type of work. I dont think its appropriate for this though. QML tool i use is too complex and make for project where a very high level of control over the rendering is needed.

**Comment by kardianos** [comment]
@gedw99 Right now I am developing a new package that will allow for different visualizations in a branch. For the command line tool I plan to keep it console based. For the default command line tool I plan to do the following:
vendor list:
x y local-path [canonical-path]
x and y being status and coded location.

Then either a different command or a flag, I will have a similar listing but with also showing what imports that package.
x y local-path [canonical-path] imported-by-list

Then have the ability to show that from a particular node's (package) perspective only as well.

I'll need to test how that works with large code bases like k8s and cockroachdb. If you have other ideas for displaying information on a terminal, I'd love to hear it. Also the structure is such that it will be easy to import a package and make a different visualization in QML, SVG, or HTML.

**Comment by gedw99** [comment]
@kardianos  this sounds excellent. CLi first makes sense i think also.
yes, visual tooling afterwards

**Comment by kardianos** [comment]
Conflict management is ready to be added to the UI.
Adjust the code here for a starting point:
https://github.com/kardianos/govendor/blob/master/run.go#L275

**Comment by kardianos** [comment]
An interactive CLI is needed. See:
https://github.com/kardianos/govendor/blob/master/DESIGN.md
